sql-api/src/main/java/io/cdap/pipeline/sql/api/template/TableInfo.java

```diff
@@ -16,32 +16,29 @@
 
 package io.cdap.pipeline.sql.api.template;
 
-import com.google.common.collect.ImmutableList;
 import org.apache.calcite.rel.type.RelDataType;
 import org.apache.calcite.rel.type.RelDataTypeFactory;
 import org.apache.calcite.rel.type.StructKind;
 import org.apache.calcite.schema.impl.AbstractTable;
 import org.apache.calcite.sql.type.SqlTypeName;
 
 import java.util.ArrayList;
+import java.util.Collections;
 import java.util.List;
-import javax.annotation.Nullable;
 
 /**
- * Represents a Calcite Table with a schema and no rows.
+ * Represents a Calcite table with a type schema.
  */
 public class TableInfo extends AbstractTable {
-  String tableName;
-  List<String> columnNames;
-  List<SqlTypeName> columnTypes;
-  RelDataType type;
+  private final String tableName;
+  private final List<String> columnNames;
+  private final List<SqlTypeName> columnTypes;
```

Comment:
This won't work for all data types. Also, it requires the plugin to map their db specific types into SqlTypeName, which eventually gets mapped to a RelDataType. It seems more straightforward to just make the plugin responsible for going straight to the RelDataType.

Instead of setting these things with a Builder, I think it's better to make this into an abstract class that extends AbstractTable, with just the additional method to get the table name. Plugins will be responsible for implementing the class, and the BQ source and sink can share the same implementation. Also, do you also need more info like the database name?

Reply:
Moved the implementation to the BigQuery plugins folder and added a TableDelegate class to support general Calcite tables.